packages/netstring/reader.js

```diff
@@ -2,67 +2,113 @@
 /// <reference types="ses"/>
 
 const COLON = ':'.charCodeAt(0);
-
-const decoder = new TextDecoder();
+const COMMA = ','.charCodeAt(0);
+const ZERO = '0'.charCodeAt(0);
+const NINE = '9'.charCodeAt(0);
 
 /**
  * @param {Iterable<Uint8Array> | AsyncIterable<Uint8Array>} input
  * @param {Object} [opts]
  * @param {string} [opts.name]
- * @param {number} [opts.capacity]
  */
-async function* makeNetstringIterator(
-  input,
-  { name = '<unknown>', capacity = 1024 } = {},
-) {
-  let length = 0;
-  let buffer = new Uint8Array(capacity);
+async function* makeNetstringIterator(input, { name = '<unknown>' } = {}) {
+  // byte offset of data consumed so far in the input stream
   let offset = 0;
 
+  // The iterator can be in 2 states: waiting for the length, or waiting for the data
+  // - When waiting for the length, the lengthBuffer is an array containing
+  //   digits charCodes for the length prefix
+  // - When waiting for the data, the dataBuffer is either:
+  //   - null to indicate no data has been received yet.
+  //   - A newly allocated buffer large enough to accommodate the whole expected data.
+  //   In either case, remainingDataLength contains the length of the data to read.
+  //   If the whole data is received in one chunk, no copy is made.
+  /** @type {number[] | null} */
+  let lengthBuffer = [];
+  /** @type {Uint8Array | null} */
+  let dataBuffer = null;
+  let remainingDataLength = -1;
+
   for await (const chunk of input) {
-    if (length + chunk.byteLength >= capacity) {
-      while (length + chunk.byteLength >= capacity) {
-        capacity *= 2;
-      }
-      const replacement = new Uint8Array(capacity);
-      replacement.set(buffer, 0);
-      buffer = replacement;
-    }
-    buffer.set(chunk, length);
-    length += chunk.byteLength;
+    let buffer = chunk;
 
-    let drained = false;
-    while (!drained && length > 0) {
-      const colon = buffer.indexOf(COLON);
-      if (colon === 0) {
-        throw new Error(
-          `Expected number before colon at offset ${offset} of ${name}`,
-        );
-      } else if (colon > 0) {
-        const prefixBytes = buffer.subarray(0, colon);
-        const prefixString = decoder.decode(prefixBytes);
-        const contentLength = +prefixString;
-        if (Number.isNaN(contentLength)) {
-          throw new Error(
-            `Invalid netstring prefix length ${prefixString} at offset ${offset} of ${name}`,
-          );
+    while (buffer.length) {
+      // Waiting for full length prefix
+      if (lengthBuffer) {
+        let i = 0;
+        while (i < buffer.length) {
+          const c = buffer[i];
+          i += 1;
+          if (c >= ZERO && c <= NINE) {
+            lengthBuffer.push(c);
+          } else if (c === COLON && lengthBuffer.length) {
+            lengthBuffer.push(c);
+            break;
+          } else {
```

Comment:
Should throw an error of the lengthBuffer length exceeds the maximum length necessary to express the maximum message length, so the reader can’t be induced to OOM by a long length prefix.

Reply:
Similarly below. Can punt to an issue since it would be a new level of hardening unrelated to this performance push.

Reply:
What do you mean by maximum message length? I believe the netstring spec doesn't have such a notion (even though the example implies a max length of 999999999 bytes). The tagged netstring spec does have that length as specified limit.

Would we consider 1GiB-1 as the upper limit, or would it be a config option of the reader?

Reply:
We should definitely have a config option and highly encourage a reasonable limit. Protobuf has a hard limit at 2GB, which is close enough to Infinity to be a reasonable default but not adequate for limiting the sender’s ability to induce an OOMkill remotely.

Reply:
Done, made configurable with the default being what the netstring specs imply.